program/src/processor.rs

```diff
@@ -108,35 +117,58 @@ fn create_or_update_v1(
         &[bump],
     ];
 
+    let data_len = match ctx.accounts.buffer_pda_info {
+        Some(account_info) => account_info.data_len(),
+        None => serialized_rule_set.len(),
+    };
+
     // Create or allocate, resize or reallocate RuleSet PDA.
     if ctx.accounts.rule_set_pda_info.data_is_empty() {
         create_or_allocate_account_raw(
             *program_id,
             ctx.accounts.rule_set_pda_info,
             ctx.accounts.system_program_info,
             ctx.accounts.payer_info,
-            serialized_rule_set.len(),
+            data_len,
             rule_set_seeds,
         )?;
     } else {
         resize_or_reallocate_account_raw(
             ctx.accounts.rule_set_pda_info,
             ctx.accounts.payer_info,
             ctx.accounts.system_program_info,
-            serialized_rule_set.len(),
+            data_len,
         )?;
     }
 
-    // Copy user-pre-serialized RuleSet to PDA account.
-    sol_memcpy(
-        &mut ctx
-            .accounts
-            .rule_set_pda_info
-            .try_borrow_mut_data()
-            .unwrap(),
-        &serialized_rule_set,
-        serialized_rule_set.len(),
-    );
+    match ctx.accounts.buffer_pda_info {
+        Some(account_info) => {
+            msg!("Using buffer account for RuleSet serialization.");
+            // Copy user-pre-serialized RuleSet to PDA account.
+            sol_memcpy(
+                &mut ctx
+                    .accounts
+                    .rule_set_pda_info
+                    .try_borrow_mut_data()
+                    .unwrap(),
+                &account_info.data.borrow(),
+                data_len,
```

Comment:
Yeah, we could do it in one account, which is what I originally tried to do, but in complicates the creation logic. The current way allows a clone so people can use an existing rule set instead of a buffer, which I think may come in handy.